engineioxide/src/payload.rs

```diff
@@ -0,0 +1,140 @@
+use std::{io::BufRead, vec};
+
+use crate::{service::ProtocolVersion, errors::Error};
+
+pub const PACKET_SEPARATOR: u8 = b'\x1e';
+
+/// A payload is a series of encoded packets tied together.
+/// How packets are tied together depends on the protocol.
+pub struct Payload<R: BufRead> {
+    reader: R,
+    buffer: Vec<u8>,
+    #[allow(dead_code)]
+    protocol: ProtocolVersion,
+}
+
+type Item = Result<String, Error>;
+
+impl<R: BufRead> Payload<R> {
+    pub fn new(protocol: ProtocolVersion, data: R) -> Self {
+        Payload {
+            reader: data,
+            buffer: vec![],
+            protocol,
+        }
+    }
+
+    #[cfg(feature = "v3")]
+    fn next_v3(&mut self) -> Option<Item> {
+        match self.reader.read_until(b':', &mut self.buffer) {
+            Ok(bytes_read) => (bytes_read > 0).then(|| {
+                if self.buffer.ends_with(&[b':']) {
+                    self.buffer.pop();
+                }
+
+                let buffer = std::mem::take(&mut self.buffer);
+                let length = String::from_utf8(buffer)
+                    .map_err(|_| Error::InvalidPacketLength)
+                    .and_then(|s| s.parse::<usize>().map_err(|_| Error::InvalidPacketLength))?;
+                
+                self.buffer.resize(length, 0);
+                self.reader.read_exact(&mut self.buffer)?;
+
+                let buffer = std::mem::take(&mut self.buffer);
+                String::from_utf8(buffer).map_err(Into::into)
+            }),
+            Err(e) => Some(Err(Error::Io(e))),
+        }
+    }
+
+    #[cfg(feature = "v4")]
+    fn next_v4(&mut self) -> Option<Item> {
```

Comment:
Same here we can improve a lot the code by pre allocating the buffer and not resizing it at every iterations. But it could the subject of another PR.